vehicle/homeassistant.go

```diff
@@ -0,0 +1,320 @@
+//     services:
+//       start_charging: script.id4_start   # Start charging (optional)
+//       stop_charging:  script.id4_stop    # Stop charging (optional)
+//       wakeup: script.id4_wakeup          # Wake up vehicle (optional)
+//     capacity: 77
+//
+// Notes:
+// - All sensors must exist as entities in Home Assistant and provide a suitable value.
+// - Changes to the vehicle API are handled exclusively in Home Assistant scripts/sensors.
+// - Unused fields can be omitted.
+
+package vehicle
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+)
+
+// -----------------------------
+// Configuration structures
+// -----------------------------
+
+type haSensors struct {
+	Soc           string `mapstructure:"soc"`           // z.B. sensor.id4_soc  (erforderlich)
+	Range         string `mapstructure:"range"`         // optional
+	Status        string `mapstructure:"status"`        // optional
+	LimitSoc      string `mapstructure:"limitSoc"`      // optional
+	Odometer      string `mapstructure:"odometer"`      // optional
+	Climater      string `mapstructure:"climater"`      // optional
+	MaxCurrent    string `mapstructure:"maxCurrent"`    // optional
+	GetMaxCurrent string `mapstructure:"getMaxCurrent"` // optional
+	FinishTime    string `mapstructure:"finishTime"`    // optional
+}
+
+type haServices struct {
+	Start  string `mapstructure:"start_charging"` // script.*  optional
+	Stop   string `mapstructure:"stop_charging"`  // script.*  optional
+	Wakeup string `mapstructure:"wakeup"`         // script.*  optional
+}
+
+type haConfig struct {
+	embed    `mapstructure:",squash"`
+	Host     string     `mapstructure:"host"`  // http://ha:8123
+	Token    string     `mapstructure:"token"` // Long-Lived Token
+	Sensors  haSensors  `mapstructure:"sensors"`
+	Services haServices `mapstructure:"services"`
+}
+
+// -----------------------------
+// Adapter implementation
+// -----------------------------
+
+type HomeAssistant struct {
+	*embed
+	*request.Helper
+	conf         haConfig
+	startScript  string
+	stopScript   string
+	wakeupScript string
+}
+
+// Register on startup
+func init() { registry.Add("homeassistant", newHAFromConfig) }
+
+// Constructor from YAML config
+func newHAFromConfig(other map[string]any) (api.Vehicle, error) {
+	var cc haConfig
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+	switch {
+	case cc.Host == "":
+		return nil, fmt.Errorf("homeassistant: host missing")
+	case cc.Token == "":
+		return nil, fmt.Errorf("homeassistant: token missing")
+	case cc.Sensors.Soc == "":
+		return nil, fmt.Errorf("homeassistant: sensors.soc missing")
+	}
+
+	log := util.NewLogger("ha-vehicle").Redact(cc.Token)
+	return &HomeAssistant{
+		embed:        &cc.embed,
+		Helper:       request.NewHelper(log),
+		conf:         cc,
+		startScript:  cc.Services.Start,
```

Comment:
Why are both conf and conf.services copied?

Reply:
solved

Reply:
Du musst die nicht alle kommentieren ;). Wenn sich das diff ändert macht GH die automatisch zu.